include/seqan3/alignment/data_structure/scoring_scheme_base.hpp

```diff
@@ -0,0 +1,220 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides seqan3::scoring_scheme_base.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <range/v3/algorithm/copy.hpp>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/core/detail/strong_type.hpp>
+#include <seqan3/core/concept/cereal.hpp>
+
+#if SEQAN3_WITH_CEREAL
+#include <cereal/types/array.hpp>
+#endif // SEQAN3_WITH_CEREAL
+
+namespace seqan3
+{
+
+/*!\brief A strong type of underlying type `int8_t` that represents the score two matching characters.
+ * \ingroup data_structure
+ * \see scoring_scheme_base::set_scheme_simple
+ */
+struct match_score : detail::strong_type<int8_t, match_score, detail::strong_type_skill::convert>
```

Comment:
> We need them parameterized as also floating point scores will be needed.

![image](https://user-images.githubusercontent.com/2996863/44030113-499f9018-9eef-11e8-85a8-e68602e79d58.png)

Reply:
Your suggestion for solving this with type deduction guides really isn't feasible since our underlying type is `int8_t`, but `match_score{11}` resolves to `match_score<int>{11}` (and that of course is not convertible to `match_score<int8_t>` because strong types) so we would need to have our users call
```cpp
scheme1.set_simple_scheme(match_score<int8_t>{11}, mismatch_score<int8_t>{-7});
```
 which is ridiculously complex and error prone.
The same goes for deduction in the constructor.

The only thing I can think of would be to add a skill to the strong types that allows strong type of some integral type to be converted to another. But I think that is also error-prone (even if optional) and defeats the purpose of strong types.

I don't see a way to resolve this cleanly and am afraid we nee to drop strong types from the interface if the underlying type really needs to be configurable. For the concrete interface I think it would be ok since we are dealing with only two parameters and no sane person would think of writing the mismatch score before the match score. [although in general of course I would like to use strong types very much]

Reply:
First, there is no rational to enforce the underlying score to be int8_t. We had use-cases where the score is set to values beyond the value range of int8_t to express for example to never allow mismatches.  The only thing we do know is, that the specific similarity matrices don't need a higher type and int8_t is for them a good choice. So the score type should be deduced from them, which we can do.
I really think we should keep the score type dependent of the user input. And well, in 99.9% the user just doesn't care about the specific type and it will be sufficient to just assume int/double for the underlying type so they will never have to write the expression from your comment. But, in case you really want to specify the type you can use regular c++ template declaration to do so. I honestly don't see the problem with that and how this interferes with strong types.
That would also allow us to use floating point scoring matrices for amino acids, if applications might need this. I mean at the moment I only know use-cases for rna/dna related scoring schemes.

Reply:
> I really think we should keep the score type dependent of the user input. And well, in 99.9% the user just doesn't care about the specific type and it will be sufficient to just assume int/double for the underlying type so they will never have to write the expression from your comment.

No, I think you didn't get my point. The strong types in the function interface are fixed for every scheme once the scheme is defined. Unless we a do a "pre-selection" like above, you will get conflicts, because numbers always deduce to int (unless postfixed by e.g. `l`) and this is incompatible with type expected.

> That would also allow us to use floating point scoring matrices for amino acids, if applications might need this. I mean at the moment I only know use-cases for rna/dna related scoring schemes.

Whether or not aminoacids can be floats is entirely independent of strong types. I just didn't templatise it, because I didn't see the need.